File: legacy/packages/contracts/src/game-adapter.ts
import type { ArtifactRef, BoundingBox, GameId, JsonObject, Locale, TimeRange, Uuid } from "./common.js";
import type { ContentType } from "./editorial.js";

export type AdapterCapability = "game_detection" | "screen_classification" | "hud" | "menus" | "entities" | "events" | "text_normalization" | "guided_search" | "knowledge" | "pronunciation" | "narrative_templates";

export interface GameAdapterDescriptor {
  readonly id: string;
  readonly game_id: Exclude<GameId, "unknown">;
  readonly version: string;
  readonly contract_version: "1.0";
  readonly display_name: string;
  readonly supported_game_versions: readonly string[];
  readonly supported_locales: readonly Locale[];
  readonly capabilities: readonly AdapterCapability[];
  readonly knowledge_namespace: string;
}

export interface FrameRef {
  readonly artifact: ArtifactRef;
  readonly timestamp_ms: number;
  readonly width: number;
  readonly height: number;
}

export interface SegmentRef {
  readonly segment_id: Uuid;
  readonly media_id: Uuid;
  readonly range: TimeRange;
  readonly representative_frames: readonly FrameRef[];
}

export interface Detection {
  readonly label: string;
  readonly confidence: number;
  readonly region: BoundingBox | null;
  readonly range: TimeRange | null;
  readonly detector_version: string;
  readonly attributes: JsonObject;
}

export interface EntityDetection extends Detection {
  readonly entity_type: string;
  readonly canonical_id: string | null;
}

export interface GameEvent extends Detection {
  readonly event_type: string;
  readonly entity_ids: readonly string[];
}

export interface NarrativeTemplate {
  readonly id: string;
  readonly version: string;
  readonly content_type: ContentType;
  readonly required_beats: readonly string[];
  readonly optional_beats: readonly string[];
}

export interface GameEntity {
  readonly canonical_id: string;
  readonly entity_type: string;
  readonly display_name: string;
  readonly aliases: readonly string[];
  readonly confidence: number;
}

export interface GameAdapter {
  readonly descriptor: GameAdapterDescriptor;
  detect_game(frame: FrameRef): Promise<Detection>;
  detect_hud(frame: FrameRef): Promise<readonly Detection[]>;
  detect_menus(frame: FrameRef): Promise<readonly Detection[]>;
  detect_entities(frame: FrameRef): Promise<readonly EntityDetection[]>;
  detect_events(segment: SegmentRef): Promise<readonly GameEvent[]>;
  normalize_text(text: string, locale: Locale): string;
  resolve_entity(text: string, context: JsonObject): Promise<GameEntity | null>;
  get_content_templates(content_type: ContentType): readonly NarrativeTemplate[];
  get_expected_events(content_type: ContentType): readonly string[];
  get_knowledge_namespace(): string;
  get_pronunciation_lexicon(locale: Locale): Readonly<Record<string, string>>;
}

File: legacy/packages/contracts/python/src/gta_ai_contracts/game_adapter.py
from __future__ import annotations

from typing import Literal, Protocol
from uuid import UUID

from pydantic import Field

from .common import ArtifactRef, BoundingBox, JsonValue, StudioModel, TimeRange, UnitScore


class GameAdapterDescriptor(StudioModel):
    id: str
    game_id: Literal["gta5", "gta6"]
    version: str
    contract_version: Literal["1.0"] = "1.0"
    display_name: str
    supported_game_versions: list[str] = Field(default_factory=list)
    supported_locales: list[str] = Field(default_factory=list)
    capabilities: list[str] = Field(default_factory=list)
    knowledge_namespace: str


class Detection(StudioModel):
    label: str
    confidence: UnitScore
    region: BoundingBox | None = None
    range: TimeRange | None = None
    detector_version: str
    attributes: dict[str, JsonValue] = Field(default_factory=dict)


class FrameRef(StudioModel):
    artifact: ArtifactRef
    timestamp_ms: int = Field(ge=0)
    width: int = Field(gt=0)
    height: int = Field(gt=0)


class SegmentRef(StudioModel):
    segment_id: UUID
    media_id: UUID
    range: TimeRange
    representative_frames: list[FrameRef] = Field(default_factory=list)


class EntityDetection(Detection):
    entity_type: str
    canonical_id: str | None = None


class GameEvent(Detection):
    event_type: str
    entity_ids: list[str] = Field(default_factory=list)


class NarrativeTemplate(StudioModel):
    id: str
    version: str
    content_type: str
    required_beats: list[str] = Field(default_factory=list)
    optional_beats: list[str] = Field(default_factory=list)


class GameEntity(StudioModel):
    canonical_id: str
    entity_type: str
    display_name: str
    aliases: list[str] = Field(default_factory=list)
    confidence: UnitScore


class GameAdapter(Protocol):
    descriptor: GameAdapterDescriptor

    async def detect_game(self, frame: FrameRef) -> Detection: ...
    async def detect_hud(self, frame: FrameRef) -> list[Detection]: ...
    async def detect_menus(self, frame: FrameRef) -> list[Detection]: ...
    async def detect_entities(self, frame: FrameRef) -> list[EntityDetection]: ...
    async def detect_events(self, segment: SegmentRef) -> list[GameEvent]: ...
    def normalize_text(self, text: str, locale: str) -> str: ...
    async def resolve_entity(self, text: str, context: dict[str, JsonValue]) -> GameEntity | None: ...
    def get_content_templates(self, content_type: str) -> list[NarrativeTemplate]: ...
    def get_expected_events(self, content_type: str) -> list[str]: ...
    def get_knowledge_namespace(self) -> str: ...
    def get_pronunciation_lexicon(self, locale: str) -> dict[str, str]: ...
